src/modules/rlm_cache/drivers/rlm_cache_redis/rlm_cache_redis.c

```diff
@@ -0,0 +1,417 @@
+				ERROR("rlm_redis (%s): Failed authentication: reply %s",
+				       inst->name, reply->str);
+				goto do_close;
+			}
+			break;	/* else it's OK */
+
+		default:
+			ERROR("rlm_redis (%s): Unexpected reply to AUTH",
+			       inst->name);
+			goto do_close;
+		}
+
+		freeReplyObject(reply);
+	}
+
+	randle = talloc_zero(ctx, rlm_cache_redis_handle_t);
+	randle->conn = conn;
+	talloc_set_destructor(randle, _mod_conn_free);
+
+	return randle;
+}
+
+/** Cleanup a rlm_cache_redis instance
+ *
+ * @param driver to free.
+ * @return 0
+ */
+static int _mod_detach(rlm_cache_redis_t *driver)
+{
+	fr_connection_pool_free(driver->pool);
+	return 0;
+}
+
+/** Create a new rlm_cache_redis instance
+ *
+ * @param conf memcached specific conf section.
+ * @param inst main rlm_cache instance.
+ * @return 0 on success, -1 on failure.
+ */
+static int mod_instantiate(CONF_SECTION *conf, rlm_cache_t *inst)
+{
+	rlm_cache_redis_t *driver;
+	char buffer[256];
+	static bool version_done;
+
+	buffer[0] = '\0';
+
+	/*
+	 *	Get version info from the libmemcached API.
+	 */
+	if (!version_done) {
+		version_done = true;
+		INFO("rlm_cache_redis: libhires version: %d.%d.%d", HIREDIS_MAJOR, HIREDIS_MINOR, HIREDIS_PATCH);
+	}
+
+	driver = talloc_zero(inst, rlm_cache_redis_t);
+	talloc_set_destructor(driver, _mod_detach);
+	if (cf_section_parse(conf, driver, driver_config) < 0) return -1;
+
+	inst->driver = driver;
+
+	snprintf(buffer, sizeof(buffer), "rlm_cache (%s)", inst->name);
+
+	driver->pool = fr_connection_pool_module_init(conf, inst, mod_conn_create, NULL, buffer);
+	if (!driver->pool) return -1;
+
+	if (inst->max_entries > 0) WARN("rlm_cache_redis: max_entries is not supported by this driver");
+
+	return 0;
+}
+
+static void cache_entry_free(rlm_cache_entry_t *c)
+{
+	talloc_free(c);
+}
+
+/** Locate a cache entry in memcached
+ *
+ * @param out Where to write the pointer to the cach entry.
+ * @param inst main rlm_cache instance.
+ * @param request The current request.
+ * @param handle Pointer to memcached handle.
+ * @param key to search for.
+ * @return CACHE_OK on success CACHE_MISS if no entry found, CACHE_ERROR on error.
+ */
+static cache_status_t cache_entry_find(rlm_cache_entry_t **out, UNUSED rlm_cache_t *inst, REQUEST *request,
+				       rlm_cache_handle_t **handle, char const *key)
+{
+	rlm_cache_redis_handle_t *randle = *handle;
+	redisReply *reply;
+	rlm_cache_entry_t *c;
+	int ret;
+
+	reply = redisCommand(randle->conn,"GET %s", key);
+	if (!reply) {
+		RERROR("Failed retrieving entry for key \"%s\"", key);
+		return CACHE_ERROR;
+	}
+
+	c = talloc_zero(NULL,  rlm_cache_entry_t);
+	switch (reply->type) {
+	case REDIS_REPLY_STRING:
+		ret = cache_deserialize(c, reply->str, reply->len);
+		if (ret < 0) {
+			RERROR("%s", fr_strerror());
+		error:
+			talloc_free(c);
+			freeReplyObject(reply);
+			return CACHE_ERROR;
+		}
+		break;
+	case REDIS_REPLY_NIL:
+		talloc_free(c);
+		freeReplyObject(reply);
+		return CACHE_MISS;
```

Comment:
Maybe a simple `goto error;` here could be helpful.

Reply:
But it's returning notfound... which is is correct